### PR TITLE
Refactor: Modularise Validate-ServiceReference.ps1 and relocate to tests/

### LINKED ISSUE #116: Refactor: Break up monolithic Validate-ServiceReference.ps1 for maintainability
## Problem

`Validate-ServiceReference.ps1` has grown to **809 lines** with a **534-line god function** (`Test-ServiceReference`) containing **23+ inline validation checks**. This monolithic structure creates significant maintainability challenges:

- **AI agent context window pressure** — agents must load the entire 809-line file to safely modify any single validation rule, consuming context budget and increasing error risk.
- **God function** — `Test-ServiceReference` (lines 158–692) handles structure, style, and content validation in a single function, violating single-responsibility.
- **Repeated boilerplate** — 20+ identical check-result creation patterns (`@{ Name=...; Pass=...; Message=... }`) with no shared factory.
- **Magic values** — hardcoded numbers (`45`, `100`), category lists, and regex patterns scattered throughout with no centralised config.
- **Mixed concerns** — structural validation (YAML, sections, ordering), style rules (trap format, warnings), and content rules (scaling parameters, serviceName consistency) are interleaved rather than grouped.
- **No traceability** — validation rules have no comments linking back to TEMPLATE.md or CONTRIBUTING.md requirements.

## Proposed Solution

Adopt a **hybrid refactoring approach** that splits simple declarative rules from complex procedural logic:

1. **Extract a declarative config block** (~50 lines) for simple rules — pattern matches, field presence checks, line limits, and valid category/section lists. Centralises all magic values.

2. **Split `Test-ServiceReference` into focused validator functions** grouped by concern:
   - `Test-FrontMatter` — YAML presence, required fields, category, file placement
   - `Test-DocumentStructure` — required sections, section ordering, optional section placement
   - `Test-StyleCompliance` — verified dates, trap format, warning format, hardcoded prices
   - `Test-ContentRules` — 45-line rule, scaling parameters, serviceName consistency, template comments

3. **Extract shared helpers:**
   - `New-ValidationCheck` — factory to eliminate 20+ repeated result-object boilerplate
   - `Get-NonCodeLines` — reusable line filter (used 3+ times)
   - `Get-H2Sections` — reusable section parser

4. **Slim orchestrator** — `Test-ServiceReference` becomes a ~100-line function that calls the focused validators and aggregates results.

### Expected Outcome

- Any single validation concern fits in **~150–200 lines** instead of 809, reducing AI agent context usage by ~70%.
- Adding a new validation rule means editing a single focused function or adding a config entry.
- Existing validation behaviour is preserved — this is a structural refactor, not a rule change.

> **Note:** This proposal is pending verification. Implementation plan to be confirmed before work begins.

### PR DESCRIPTION
**Pull request by ahmadabdalla**
## Summary

Breaks up the 809-line monolithic `Validate-ServiceReference.ps1` into 10 focused modules and relocates CI tooling from `skills/azure-cost-calculator/scripts/` to `tests/` at the repo root.

Closes #116

### Modularisation (from 1 file → 11 files)

| File | Lines | Purpose |
|------|-------|---------|
| `tests/Validate-ServiceReference.ps1` | 125 | Slim orchestrator |
| `tests/lib/validation/ValidationConfig.psd1` | 18 | Centralised config (categories, sections, limits) |
| `tests/lib/validation/New-ValidationCheck.ps1` | 47 | Factory for check result objects |
| `tests/lib/validation/Get-FrontMatter.ps1` | 104 | YAML front matter parser |
| `tests/lib/validation/Get-H2Section.ps1` | 34 | H2 heading parser |
| `tests/lib/validation/Get-NonCodeLine.ps1` | 40 | Code-block-aware line filter |
| `tests/lib/validation/Test-FrontMatter.ps1` | 93 | YAML, fields, aliases, category, placement (6 checks) |
| `tests/lib/validation/Test-DocumentStructure.ps1` | 91 | Required sections, ordering (5 checks) |
| `tests/lib/validation/Test-StyleCompliance.ps1` | 81 | Trap format, verified dates, warnings (5 checks) |
| `tests/lib/validation/Test-ContentRule.ps1` | 232 | 45-line rule, scaling, serviceName consistency (9 checks) |
| `tests/lib/validation/Test-AliasUniqueness.ps1` | 80 | Cross-file alias collision (1 check) |

### Directory relocation
- **Before**: `skills/azure-cost-calculator/scripts/Validate-ServiceReference.ps1` (mixed with end-user scripts)
- **After**: `tests/Validate-ServiceReference.ps1` (repo root — CI tooling separated from skill content)

### Code quality improvements
- 100% PSScriptAnalyzer compliance (zero warnings/errors)
- `#requires -Version 7.0` for explicit PS7 dependency
- Consistent comma-trick for List returns (prevents pipeline unrolling)
- `New-ValidationCheck` factory eliminates 20+ repeated boilerplate blocks
- All magic values centralised in `ValidationConfig.psd1`
- `-Encoding UTF8` on all `Get-Content` calls
- `ToLowerInvariant()` for deterministic alias comparison
- Safe `Split-Path` for filename extraction (no array index throws)
- `Write-Warning` on empty wildcard matches
- Fixed pre-existing StrictMode crash on single-line and empty-wildcard inputs

### References updated
- `CONTRIBUTING.md`, `AGENTS.md`, `.github/workflows/validate-service-references.yml`, `.github/PULL_REQUEST_TEMPLATE.md`

## Verification

- **Regression**: 1,483 check output lines across all service files — zero diff from original
- **PSScriptAnalyzer**: zero issues
- **CI simulation**: 9 test scenarios passing (valid files, multi-file PR, nonexistent files, empty wildcards, malformed files, TEMPLATE.md, no-flag mode)
- **Edge cases**: StrictMode compliance verified for single-line files, empty YAML scalars, bare filenames

## Checklist

- [x] All 23 validation checks produce identical names, pass/fail, and messages
- [x] PSScriptAnalyzer: zero warnings/errors
- [x] CI workflow path updated
- [x] Documentation paths updated
- [x] Default `ServicesRoot` works without explicit flag
- [x] No breaking changes to CLI interface

### CODE REVIEW

File: tests/lib/validation/Get-FrontMatter.ps1 [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                if ($nextLine -match '^\s+\-\s*(.+)$') {
+                    $items += $Matches[1].Trim()
+                    $index++
+                    continue
+                }
+                break
+            }
+            # Normalise to bracket string so downstream sees the same format as inline aliases
+            if ($items.Count -eq 0) {
+                $result.Fields[$key] = ''
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply tests/lib/validation/Get-FrontMatter.ps1]: Good catch — the `if/else` was a deviation from the original script, which unconditionally does `"[" + ($items -join ", ") + "]"`. When `$items` is empty, `-join` naturally produces an empty string, so the result is `"[]"` — matching the original behavior.

Rather than just changing `""` to `"[]"`, I've collapsed the branch entirely back to the original one-liner in 7acb883. This is cleaner and preserves the zero-diff claim.
